src/extract/global_greedy_dag.rs

```diff
@@ -0,0 +1,193 @@
+use rpds::{HashTrieMap, HashTrieSet};
+
+use super::*;
+
+type TermId = usize;
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+struct Term {
+    op: String,
+    children: Vec<TermId>,
+}
+
+type Reachable = HashTrieSet<ClassId>;
+
+struct TermInfo {
+    node: NodeId,
+    eclass: ClassId,
+    node_cost: NotNan<f64>,
+    total_cost: NotNan<f64>,
+    // store the set of reachable terms from this term
+    reachable: Reachable,
+    size: usize,
+}
+
+// A TermDag needs to store terms that share common
+// subterms using a hashmap.
+// However, it also critically needs to be able to answer
+// reachability queries in this dag `reachable`.
+// This prevents double-counting costs when
+// computing the cost of a term.
+#[derive(Default)]
+pub struct TermDag {
+    nodes: Vec<Term>,
+    info: Vec<TermInfo>,
+    hash_cons: HashMap<Term, TermId>,
+}
+
+impl TermDag {
+    // Makes a new term using a node and children terms
+    // Correctly computes total_cost with sharing
+    // If this term contains itself, returns None
+    // If this term costs more than target, returns None
+    pub fn make(
+        &mut self,
+        node_id: NodeId,
+        node: &Node,
+        children: Vec<TermId>,
+        target: NotNan<f64>,
+    ) -> Option<TermId> {
+        let term = Term {
+            op: node.op.clone(),
+            children: children.clone(),
+        };
+
+        if let Some(id) = self.hash_cons.get(&term) {
+            return Some(*id);
+        }
+
+        let node_cost = node.cost;
+
+        if children.is_empty() {
+            let next_id = self.nodes.len();
+            self.nodes.push(term.clone());
+            self.info.push(TermInfo {
+                node: node_id,
+                eclass: node.eclass.clone(),
+                node_cost,
+                total_cost: node_cost,
+                reachable: [node.eclass.clone()].into_iter().collect(),
+                size: 1,
+            });
+            self.hash_cons.insert(term, next_id);
+            Some(next_id)
+        } else {
+            // check if children contains this node
+            for child in &children {
+                if self.info[*child].reachable.contains(&node.eclass) {
+                    return None;
+                }
+            }
+
+            let biggest_child = (0..children.len())
+                .max_by_key(|i| self.info[children[*i]].size)
+                .unwrap();
+
+            let mut cost = node_cost + self.total_cost(children[biggest_child]);
+            let mut reachable = Box::new(self.info[children[biggest_child]].reachable.clone());
```

Comment:
Why `Box`?

Reply:
This is my hack to keep track of a persistent set which we are "mutating" during recursive calls. We can't mutate the container itself, so we need a wrapper as far as I know.
Is there a preferred solution in rust?

Alternatively, the recursive function could also return the new set each time.

Reply:
* I'm not against having subcalls return a new set.
* I still don't understand why you need a box here. Why can't you pass a `&mut Reachable` and then reassign as you are doing now?

Reply:
Wow TIL that `&mut` literally means pass by reference. I thought `&mut` was a mutable reference but you couldn't assign to it.

Reply:
Updated the code